### DISCUSSION TITLE
Choose one: should the `State` use `s.dot_notation` or `s["dictionary notation"]`?

### DISCUSSION THREAD

**Discussion by hollandjg** [AutoResearch/autora#575] [Core Contributions] [answered]
**Which notation do people prefer for the `State` object?**
- As dictionaries, which expose their values like `s["dictionary notation"]`
- As dataclasses which expose their values like `s.dot_notation`

We want to settle on one of these for the following reasons:
- Maintenance workload
- Clarity of having one canonical way to access values in state objects

All the features we _currently_ want on the `State` are available on both.

Differences:
- Allowed field names / keys:
  - There is no restriction on the keys which can be added to a dictionary.
  - There are some restrictions on the names of the fields in the dataclass – they can't be the same as method names we use now (`update`, `add_field`) or want to use in the future.
- Familiarity
  - Everyone familiar with python knows dictionaries well
  - Dataclasses are newer and less well understood
- Ease of adding keys:
  -  Keys can be added to a dictionary at will
  - Keys need to be explicitly added to a dataclass
- Mutability
  - Dictionaries are mutable by default
  - Dataclasses can be made immutable (and we do this on the State dataclasses)

**Comment by chadcwilliams** [comment, marked as answer]
I am voting for dataclass. 

This is a good video demonstrating and comparing both and it does a good job showing how much cleaner it is to use dataclasses: https://youtu.be/8SCYteAlqko?si=2mE_kil0skzKJlND

Although dictionaries are simpler to use, and will facilitate users with less experience, dataclasses have some extra functionality, which will facilitate more experienced users. Although it's nice to support less experienced users, I suspect AutoRA will more heavily be used by experienced users so I would focus on that. Dataclasses also have extra functionality, which may come in handy as AutoRA evolves to be more complex, or as users use it in more complex ways. 

Lastly, our experimentalists and experiment runners are functions and our theorists are classes, so choosing dataclasses over dictionaries minimizes the different types of components.

**Comment by hollandjg** [comment]
Postscripts:
- @musslick: `s.dot_notation` allows for tab-completion, autocompletion - better overview of the keys.
- @younesStrittmatter : if we want the dynamic behavior, then the dictionary is the right answer, because that is what the dictionary is designed to do. Dynamically creating a class/type doesn't feel straightforward.

**Comment by musslick** [comment]
- add metadata in the field instead of in the data class / dictionary

**Comment by chadcwilliams** [comment]
Adding a note here on our decision so this does not get lost. For the next release, we have decided to stick with the dataclass using the StandardState.
